player/lua/console.lua

```diff
@@ -1826,4 +1851,51 @@ mp.register_event('log-message', function(e)
             terminal_styles[e.level])
 end)
 
+mp.register_event('shutdown', function ()
+    if opts.persist_history == false or history_is_dirty == false then
+        return
+    end
+
+    history = histories[default_id]
+    local history_to_save = table.concat(history, '\n') .. '\n'
+    local history_path = mp.command_native({'expand-path', opts.history_path})
+
+    -- Reread the history file to get lines that other mpv instances may have
+    -- added.
+    local history_file = io.open(history_path)
+
+    if history_file then
+        for history_line in history_file:lines() do
+            local found = false
+
+            for _, history_line2 in ipairs(history) do
+                if history_line == history_line2 then
+                    found = true
+                    break
+                end
+            end
+
+            if found == false then
+                history_to_save = history_to_save .. history_line .. '\n'
+            end
+        end
+
+        history_file:close()
+    end
+
+    local error
+    history_file, error = io.open(history_path, 'w')
```

Comment:
Why not just append and avoid all the read logic?

Reply:
We don't know whether history commands are new or were read from the history file on startup. Even if we kept track of it that would break history deduplication.

Reply:
Why history deduplication is needed?

Reply:
To avoid duplicate lines? We even have an option for it.

Reply:
Well, bit too bloated for my taste, but whatever.